File: sdk-java/dephealth-core/src/main/java/biz/kryukov/dev/dephealth/parser/ConfigParser.java
package biz.kryukov.dev.dephealth.parser;

import biz.kryukov.dev.dephealth.ConfigurationException;
import biz.kryukov.dev.dephealth.DefaultPorts;
import biz.kryukov.dev.dephealth.DependencyType;
import biz.kryukov.dev.dephealth.Endpoint;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Parser for URLs, JDBC URLs, connection strings, and explicit host/port parameters.
 */
public final class ConfigParser {

    private ConfigParser() {}

    private static final Map<String, DependencyType> SCHEME_TO_TYPE = Map.ofEntries(
            Map.entry("postgres", DependencyType.POSTGRES),
            Map.entry("postgresql", DependencyType.POSTGRES),
            Map.entry("mysql", DependencyType.MYSQL),
            Map.entry("redis", DependencyType.REDIS),
            Map.entry("rediss", DependencyType.REDIS),
            Map.entry("amqp", DependencyType.AMQP),
            Map.entry("amqps", DependencyType.AMQP),
            Map.entry("http", DependencyType.HTTP),
            Map.entry("https", DependencyType.HTTP),
            Map.entry("grpc", DependencyType.GRPC),
            Map.entry("kafka", DependencyType.KAFKA)
    );

    private static final Map<String, DependencyType> JDBC_SUBPROTOCOL_TO_TYPE = Map.of(
            "postgresql", DependencyType.POSTGRES,
            "mysql", DependencyType.MYSQL
    );

    private static final Set<String> HOST_KEYS = Set.of(
            "host", "server", "data source", "address", "addr", "network address"
    );

    private static final Set<String> PORT_KEYS = Set.of("port");

    /**
     * Parses a URL (supports multi-host for Kafka, IPv6).
     *
     * @param rawUrl URL to parse
     * @return list of parsed connections (one or more for multi-host)
     */
    public static List<ParsedConnection> parseUrl(String rawUrl) {
        if (rawUrl == null || rawUrl.isBlank()) {
            throw new ConfigurationException("URL must not be empty");
        }

        // Determine the scheme
        int colonSlashSlash = rawUrl.indexOf("://");
        if (colonSlashSlash < 0) {
            throw new ConfigurationException("URL must have a scheme (e.g. http://): " + rawUrl);
        }
        String scheme = rawUrl.substring(0, colonSlashSlash).toLowerCase();
        DependencyType type = SCHEME_TO_TYPE.get(scheme);
        if (type == null) {
            throw new ConfigurationException("Unsupported URL scheme: " + scheme);
        }

        String rest = rawUrl.substring(colonSlashSlash + 3);

        // Strip userinfo (user:pass@)
        int atSign = rest.indexOf('@');
        if (atSign >= 0) {
            rest = rest.substring(atSign + 1);
        }

        // Strip path/query/fragment
        int pathStart = rest.indexOf('/');
        String hostPortPart = pathStart >= 0 ? rest.substring(0, pathStart) : rest;
        // Also strip query without path
        int queryStart = hostPortPart.indexOf('?');
        if (queryStart >= 0) {
            hostPortPart = hostPortPart.substring(0, queryStart);
        }

        // Multi-host (Kafka): host1:port1,host2:port2
        if (type == DependencyType.KAFKA || hostPortPart.contains(",")) {
            return parseMultiHost(hostPortPart, scheme, type);
        }

        String[] hp = extractHostPort(hostPortPart, scheme);
        validatePort(hp[1], rawUrl);
        return List.of(new ParsedConnection(hp[0], hp[1], type));
    }

    /**
     * Parses a JDBC URL: jdbc:subprotocol://host:port/db
     */
    public static List<ParsedConnection> parseJdbc(String jdbcUrl) {
        if (jdbcUrl == null || jdbcUrl.isBlank()) {
            throw new ConfigurationException("JDBC URL must not be empty");
        }
        if (!jdbcUrl.toLowerCase().startsWith("jdbc:")) {
            throw new ConfigurationException("JDBC URL must start with 'jdbc:': " + jdbcUrl);
        }

        String withoutJdbc = jdbcUrl.substring(5); // strip "jdbc:"

        // Determine the subprotocol
        int colonSlashSlash = withoutJdbc.indexOf("://");
        if (colonSlashSlash < 0) {
            throw new ConfigurationException("JDBC URL must contain ://: " + jdbcUrl);
        }
        String subprotocol = withoutJdbc.substring(0, colonSlashSlash).toLowerCase();
        DependencyType type = JDBC_SUBPROTOCOL_TO_TYPE.get(subprotocol);
        if (type == null) {
            throw new ConfigurationException("Unsupported JDBC subprotocol: " + subprotocol);
        }

        // Parse as a regular URL
        String normalizedUrl = subprotocol + withoutJdbc.substring(colonSlashSlash);
        return parseUrl(normalizedUrl);
    }

    /**
     * Parses a connection string in key=value;key=value format.
     *
     * @return Endpoint with host and port
     */
    public static Endpoint parseConnectionString(String connStr) {
        if (connStr == null || connStr.isBlank()) {
            throw new ConfigurationException("Connection string must not be empty");
        }

        Map<String, String> pairs = parseKeyValuePairs(connStr);

        String host = null;
        String port = null;

        for (Map.Entry<String, String> entry : pairs.entrySet()) {
            String key = entry.getKey().toLowerCase();
            String value = entry.getValue();

            if (HOST_KEYS.contains(key)) {
                // host may contain port via comma (SQL Server) or colon
                if (value.contains(",")) {
                    // SQL Server: Server=host,port
                    String[] parts = value.split(",", 2);
                    host = parts[0].trim();
                    if (port == null) {
                        port = parts[1].trim();
                    }
                } else if (value.contains(":") && !value.startsWith("[")) {
                    // host:port (but not IPv6)
                    String[] parts = value.split(":", 2);
                    host = parts[0].trim();
                    if (port == null) {
                        port = parts[1].trim();
                    }
                } else {
                    host = value.trim();
                }
            } else if (PORT_KEYS.contains(key)) {
                port = value.trim();
            }
        }

        if (host == null || host.isEmpty()) {
            throw new ConfigurationException("Connection string must contain a host key: " + connStr);
        }
        if (port == null || port.isEmpty()) {
            throw new ConfigurationException("Connection string must contain a port: " + connStr);
        }

        validatePort(port, connStr);
        return new Endpoint(host, port);
    }

    /**
     * Creates an Endpoint from explicit host and port parameters.
     */
    public static Endpoint parseParams(String host, String port) {
        if (host == null || host.isBlank()) {
            throw new ConfigurationException("host must not be empty");
        }
        if (port == null || port.isBlank()) {
            throw new ConfigurationException("port must not be empty");
        }

        // Strip IPv6 brackets
        String cleanHost = host;
        if (cleanHost.startsWith("[") && cleanHost.endsWith("]")) {
            cleanHost = cleanHost.substring(1, cleanHost.length() - 1);
        }

        validatePort(port, host + ":" + port);
        return new Endpoint(cleanHost, port);
    }

    private static List<ParsedConnection> parseMultiHost(String hostPortPart, String scheme,
                                                         DependencyType type) {
        String[] segments = hostPortPart.split(",");
        List<ParsedConnection> result = new ArrayList<>(segments.length);
        for (String segment : segments) {
            String trimmed = segment.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            String[] hp = extractHostPort(trimmed, scheme);
            validatePort(hp[1], trimmed);
            result.add(new ParsedConnection(hp[0], hp[1], type));
        }
        if (result.isEmpty()) {
            throw new ConfigurationException("No hosts found in multi-host URL: " + hostPortPart);
        }
        return result;
    }

    private static String[] extractHostPort(String hostPort, String scheme) {
        String host;
        String port;

        // IPv6: [::1]:port
        if (hostPort.startsWith("[")) {
            int closeBracket = hostPort.indexOf(']');
            if (closeBracket < 0) {
                throw new ConfigurationException("Invalid IPv6 address: " + hostPort);
            }
            host = hostPort.substring(1, closeBracket);
            String remainder = hostPort.substring(closeBracket + 1);
            if (remainder.startsWith(":")) {
                port = remainder.substring(1);
            } else {
                port = DefaultPorts.forScheme(scheme);
            }
        } else {
            int lastColon = hostPort.lastIndexOf(':');
            if (lastColon > 0 && lastColon < hostPort.length() - 1) {
                host = hostPort.substring(0, lastColon);
                port = hostPort.substring(lastColon + 1);
            } else {
                host = lastColon == hostPort.length() - 1
                        ? hostPort.substring(0, lastColon) : hostPort;
                port = DefaultPorts.forScheme(scheme);
            }
        }

        if (host == null || host.isEmpty()) {
            throw new ConfigurationException("Empty host in: " + hostPort);
        }
        if (port == null) {
            throw new ConfigurationException("Cannot determine port for: " + hostPort);
        }

        return new String[]{host, port};
    }

    private static void validatePort(String portStr, String source) {
        try {
            int port = Integer.parseInt(portStr);
            if (port < 1 || port > 65535) {
                throw new ConfigurationException(
                        "Port out of range (1-65535): " + port + " in " + source);
            }
        } catch (NumberFormatException e) {
            throw new ConfigurationException("Invalid port: '" + portStr + "' in " + source);
        }
    }

    private static Map<String, String> parseKeyValuePairs(String connStr) {
        Map<String, String> result = new LinkedHashMap<>();
        String[] pairs = connStr.split(";");
        for (String pair : pairs) {
            String trimmed = pair.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            int eq = trimmed.indexOf('=');
            if (eq < 0) {
                continue;
            }
            String key = trimmed.substring(0, eq).trim();
            String value = trimmed.substring(eq + 1).trim();
            if (!key.isEmpty()) {
                result.put(key, value);
            }
        }
        return result;
    }
}

File: sdk-csharp/DepHealth.Core/ConfigParser.cs
namespace DepHealth;

/// <summary>
/// Parser for URLs, JDBC URLs, connection strings, and explicit parameters (host/port).
/// </summary>
public static class ConfigParser
{
    private static readonly Dictionary<string, DependencyType> SchemeToType =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["postgres"] = DependencyType.Postgres,
            ["postgresql"] = DependencyType.Postgres,
            ["mysql"] = DependencyType.MySql,
            ["redis"] = DependencyType.Redis,
            ["rediss"] = DependencyType.Redis,
            ["amqp"] = DependencyType.Amqp,
            ["amqps"] = DependencyType.Amqp,
            ["http"] = DependencyType.Http,
            ["https"] = DependencyType.Http,
            ["grpc"] = DependencyType.Grpc,
            ["kafka"] = DependencyType.Kafka,
            ["ldap"] = DependencyType.Ldap,
            ["ldaps"] = DependencyType.Ldap
        };

    private static readonly Dictionary<string, DependencyType> JdbcSubprotocolToType =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["postgresql"] = DependencyType.Postgres,
            ["mysql"] = DependencyType.MySql
        };

    private static readonly HashSet<string> HostKeys =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "host", "server", "data source", "address", "addr", "network address"
        };

    private static readonly HashSet<string> PortKeys =
        new(StringComparer.OrdinalIgnoreCase) { "port" };

    /// <summary>
    /// Parses a URL (supports multi-host for Kafka, IPv6).
    /// </summary>
    public static List<ParsedConnection> ParseUrl(string rawUrl)
    {
        if (string.IsNullOrWhiteSpace(rawUrl))
        {
            throw new ConfigurationException("URL must not be empty");
        }

        var colonSlashSlash = rawUrl.IndexOf("://", StringComparison.Ordinal);
        if (colonSlashSlash < 0)
        {
            throw new ConfigurationException($"URL must have a scheme (e.g. http://): {rawUrl}");
        }

        var scheme = rawUrl[..colonSlashSlash].ToLowerInvariant();
        if (!SchemeToType.TryGetValue(scheme, out var type))
        {
            throw new ConfigurationException($"Unsupported URL scheme: {scheme}");
        }

        var rest = rawUrl[(colonSlashSlash + 3)..];

        // Strip userinfo (user:pass@)
        var atSign = rest.IndexOf('@');
        if (atSign >= 0)
        {
            rest = rest[(atSign + 1)..];
        }

        // Strip path/query/fragment
        var pathStart = rest.IndexOf('/');
        var hostPortPart = pathStart >= 0 ? rest[..pathStart] : rest;
        var queryStart = hostPortPart.IndexOf('?');
        if (queryStart >= 0)
        {
            hostPortPart = hostPortPart[..queryStart];
        }

        // Multi-host (Kafka): host1:port1,host2:port2
        if (type == DependencyType.Kafka || hostPortPart.Contains(','))
        {
            return ParseMultiHost(hostPortPart, scheme, type);
        }

        var (host, port) = ExtractHostPort(hostPortPart, scheme);
        ValidatePort(port, rawUrl);
        return [new ParsedConnection(host, port, type)];
    }

    /// <summary>
    /// Parses a JDBC URL: jdbc:subprotocol://host:port/db
    /// </summary>
    public static List<ParsedConnection> ParseJdbc(string jdbcUrl)
    {
        if (string.IsNullOrWhiteSpace(jdbcUrl))
        {
            throw new ConfigurationException("JDBC URL must not be empty");
        }

        if (!jdbcUrl.StartsWith("jdbc:", StringComparison.OrdinalIgnoreCase))
        {
            throw new ConfigurationException($"JDBC URL must start with 'jdbc:': {jdbcUrl}");
        }

        var withoutJdbc = jdbcUrl[5..];
        var colonSlashSlash = withoutJdbc.IndexOf("://", StringComparison.Ordinal);
        if (colonSlashSlash < 0)
        {
            throw new ConfigurationException($"JDBC URL must contain ://: {jdbcUrl}");
        }

        var subprotocol = withoutJdbc[..colonSlashSlash].ToLowerInvariant();
        if (!JdbcSubprotocolToType.ContainsKey(subprotocol))
        {
            throw new ConfigurationException($"Unsupported JDBC subprotocol: {subprotocol}");
        }

        var normalizedUrl = subprotocol + withoutJdbc[colonSlashSlash..];
        return ParseUrl(normalizedUrl);
    }

    /// <summary>
    /// Parses a connection string in key=value;key=value format.
    /// </summary>
    public static Endpoint ParseConnectionString(string connStr)
    {
        if (string.IsNullOrWhiteSpace(connStr))
        {
            throw new ConfigurationException("Connection string must not be empty");
        }

        var pairs = ParseKeyValuePairs(connStr);

        string? host = null;
        string? port = null;

        foreach (var (key, value) in pairs)
        {
            if (HostKeys.Contains(key))
            {
                if (value.Contains(','))
                {
                    // SQL Server: Server=host,port
                    var parts = value.Split(',', 2);
                    host = parts[0].Trim();
                    port ??= parts[1].Trim();
                }
                else if (value.Contains(':') && !value.StartsWith('['))
                {
                    // host:port (but not IPv6)
                    var parts = value.Split(':', 2);
                    host = parts[0].Trim();
                    port ??= parts[1].Trim();
                }
                else
                {
                    host = value.Trim();
                }
            }
            else if (PortKeys.Contains(key))
            {
                port = value.Trim();
            }
        }

        if (string.IsNullOrEmpty(host))
        {
            throw new ConfigurationException($"Connection string must contain a host key: {connStr}");
        }

        if (string.IsNullOrEmpty(port))
        {
            throw new ConfigurationException($"Connection string must contain a port: {connStr}");
        }

        ValidatePort(port, connStr);
        return new Endpoint(host, port);
    }

    /// <summary>
    /// Creates an Endpoint from explicit host and port parameters.
    /// </summary>
    public static Endpoint ParseParams(string host, string port)
    {
        if (string.IsNullOrWhiteSpace(host))
        {
            throw new ConfigurationException("host must not be empty");
        }

        if (string.IsNullOrWhiteSpace(port))
        {
            throw new ConfigurationException("port must not be empty");
        }

        var cleanHost = host;
        if (cleanHost.StartsWith('[') && cleanHost.EndsWith(']'))
        {
            cleanHost = cleanHost[1..^1];
        }

        ValidatePort(port, $"{host}:{port}");
        return new Endpoint(cleanHost, port);
    }

    private static List<ParsedConnection> ParseMultiHost(
        string hostPortPart, string scheme, DependencyType type)
    {
        var segments = hostPortPart.Split(',');
        var result = new List<ParsedConnection>(segments.Length);

        foreach (var segment in segments)
        {
            var trimmed = segment.Trim();
            if (string.IsNullOrEmpty(trimmed))
            {
                continue;
            }

            var (host, port) = ExtractHostPort(trimmed, scheme);
            ValidatePort(port, trimmed);
            result.Add(new ParsedConnection(host, port, type));
        }

        if (result.Count == 0)
        {
            throw new ConfigurationException($"No hosts found in multi-host URL: {hostPortPart}");
        }

        return result;
    }

    private static (string Host, string Port) ExtractHostPort(string hostPort, string scheme)
    {
        string host;
        string? port;

        // IPv6: [::1]:port
        if (hostPort.StartsWith('['))
        {
            var closeBracket = hostPort.IndexOf(']');
            if (closeBracket < 0)
            {
                throw new ConfigurationException($"Invalid IPv6 address: {hostPort}");
            }

            host = hostPort[1..closeBracket];
            var remainder = hostPort[(closeBracket + 1)..];
            port = remainder.StartsWith(':') ? remainder[1..] : DefaultPorts.ForScheme(scheme);
        }
        else
        {
            var lastColon = hostPort.LastIndexOf(':');
            if (lastColon > 0 && lastColon < hostPort.Length - 1)
            {
                host = hostPort[..lastColon];
                port = hostPort[(lastColon + 1)..];
            }
            else
            {
                host = lastColon == hostPort.Length - 1
                    ? hostPort[..lastColon]
                    : hostPort;
                port = DefaultPorts.ForScheme(scheme);
            }
        }

        if (string.IsNullOrEmpty(host))
        {
            throw new ConfigurationException($"Empty host in: {hostPort}");
        }

        if (port is null)
        {
            throw new ConfigurationException($"Cannot determine port for: {hostPort}");
        }

        return (host, port);
    }

    private static void ValidatePort(string portStr, string source)
    {
        if (!int.TryParse(portStr, out var port))
        {
            throw new ConfigurationException($"Invalid port: '{portStr}' in {source}");
        }

        if (port < 1 || port > 65535)
        {
            throw new ConfigurationException($"Port out of range (1-65535): {port} in {source}");
        }
    }

    private static Dictionary<string, string> ParseKeyValuePairs(string connStr)
    {
        var result = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var pairs = connStr.Split(';');

        foreach (var pair in pairs)
        {
            var trimmed = pair.Trim();
            if (string.IsNullOrEmpty(trimmed))
            {
                continue;
            }

            var eq = trimmed.IndexOf('=');
            if (eq < 0)
            {
                continue;
            }

            var key = trimmed[..eq].Trim();
            var value = trimmed[(eq + 1)..].Trim();
            if (!string.IsNullOrEmpty(key))
            {
                result[key] = value;
            }
        }

        return result;
    }
}
